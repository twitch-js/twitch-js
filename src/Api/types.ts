import { ApiVersions } from '../twitch'

import { FetchOptions } from '../utils/fetch'
import { LoggerOptions } from '../utils/logger'

export type ApiOptions = {
  token: string
  clientId?: string
  log?: LoggerOptions

  /**
   * Called when a 401: Unauthorized response is returned.
   * Returns a `Promise` that must resolve with the refreshed token.
   * Upon resolution, the original request (with original parameters) is repeated using the provided token.
   * @see https://github.com/twitch-js/twitch-js#refreshing-tokens
   */
  onAuthenticationFailure?: () => Promise<string>
}

export enum ApiReadyStates {
  'NOT_READY',
  'READY',
  'INITIALIZED',
}

export type ApiSettings = {
  baseUrl: string
  authorizationHeader: string
}

export const Settings: Record<ApiVersions, ApiSettings> = {
  [ApiVersions.Helix]: {
    baseUrl: 'https://api.twitch.tv/helix',
    authorizationHeader: 'Bearer',
  },
  [ApiVersions.Kraken]: {
    baseUrl: 'https://api.twitch.tv/kraken',
    authorizationHeader: 'OAuth',
  },
}

export type ApiHeaders = Record<string, string> & {
  'Client-ID'?: string
  Accept?: string
  Authorization?: string
}

type ApiVersionOptions = {
  version?: ApiVersions
}

export type ApiFetchOptions = FetchOptions & ApiVersionOptions
