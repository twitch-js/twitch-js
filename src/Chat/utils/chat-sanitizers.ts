import isEmpty from 'lodash/isEmpty'
import random from 'lodash/random'

import { ANONYMOUS_USERNAME } from '../chat-constants'

export const channel = (value?: any): string => {
  if (typeof value !== 'string' || value.length === 0) {
    return ''
  }

  value = value.toLowerCase()

  if (!value.startsWith('#')) {
    return `#${value}`
  }

  return value
}

export const token = (value: any): string => {
  if (value == null) {
    return 'TWITCHJS'
  }

  if (value.startsWith('oauth:')) {
    return value
  }

  return `oauth:${value}`
}

export const username = (value: any): string => {
  if (isEmpty(value) || value === 'justinfan') {
    return `${ANONYMOUS_USERNAME}${random(80000, 81000)}`
  }

  return value
}
