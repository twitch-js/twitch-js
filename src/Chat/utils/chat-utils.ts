import { BaseMessage, Events, Commands } from '../../twitch'

import * as constants from '../chat-constants'

export const isAuthenticationFailedMessage = (message?: BaseMessage) =>
  typeof message !== 'undefined' &&
  message.command === Commands.NOTICE &&
  message.channel === '' &&
  message.message === 'Login authentication failed'

export const getEventNameFromMessage = (message: BaseMessage) =>
  typeof message !== 'undefined' ? message.command || message.event : Events.ALL

export const isUserAnonymous = (value: string) =>
  constants.ANONYMOUS_USERNAME_RE.test(value)
