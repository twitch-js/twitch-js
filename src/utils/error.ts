import { CustomError } from 'ts-custom-error'

class TwitchJSError extends CustomError {
  timestamp: Date = new Date()

  constructor(message: string) {
    super(message)
    Object.defineProperty(this, 'name', { value: 'TwitchJSError' })
  }
}

export default TwitchJSError
