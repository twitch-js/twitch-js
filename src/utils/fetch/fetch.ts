import fetch from 'cross-fetch'
import { stringify, IStringifyOptions } from 'qs'
import camelCaseKeys from 'camelcase-keys'

import TwitchJSError from '../error'

type QueryParams = {
  /** Any query parameters you want to add to your request. */
  [key: string]: string | number | boolean
}

interface BodyParams {
  [key: string]: any
}

export type FetchOptions<Query = QueryParams, Body = BodyParams> = Omit<
  RequestInit,
  'body'
> & { search?: Query } & { body?: Body }

export class FetchError extends TwitchJSError {
  public constructor(message: string, public body?: any) {
    super(message)
    Object.defineProperty(this, 'name', { value: 'TwitchJSFetchError' })
  }
}

/**
 * Fetches URL
 */
const fetchUtil = async <
  Response = any,
  Query = QueryParams,
  Body = BodyParams
>(
  url: RequestInfo,
  options?: FetchOptions<Query, Body>,
  qsOptions?: IStringifyOptions,
) => {
  const { search, body: bodyParams, ...rest } = options || {}

  const queryParams = search
    ? stringify(search, {
        ...qsOptions,
        addQueryPrefix: true,
        arrayFormat: 'repeat',
      })
    : ''

  const jsonInit = getJsonInit(bodyParams)

  const init = {
    ...rest,
    ...jsonInit,
    headers: {
      ...options?.headers,
      ...jsonInit.headers,
    },
  }

  const response = await fetch(`${url}${queryParams}`, init)

  const json = await response.json().catch(() => ({}))

  const body = camelCaseKeys(json, { deep: true }) as
    | Response
    | { error: string; status: number; message: string }

  if ('error' in body) {
    throw new FetchError(body.message, body)
  }

  if (!response.ok) {
    throw new FetchError(response.statusText, {
      error: true,
      status: response.status,
      message: response.statusText,
    })
  }

  return body
}

const getJsonInit = (
  input?: any,
): { body?: string | FormData; headers?: HeadersInit } => {
  try {
    if (!input) {
      return {}
    }

    if (input instanceof FormData) {
      return { body: input }
    }

    const body = JSON.stringify(input)
    return { body, headers: { 'Content-Type': 'application/json' } }
  } catch (err) {
    return { body: input }
  }
}

export default fetchUtil
