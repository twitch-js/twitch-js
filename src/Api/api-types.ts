import { FetchOptions } from '../utils/fetch'
import { LoggerOptions } from '../utils/logger'

export type ApiOptions = {
  token: string
  clientId: string
  log?: LoggerOptions
  onAuthenticationFailure?: () => Promise<string>
}

export enum ApiReadyStates {
  'NOT_READY',
  'READY',
  'INITIALIZED',
}

export type ApiFetchOptions = FetchOptions
