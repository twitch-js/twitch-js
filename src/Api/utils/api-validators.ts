import invariant from 'invariant'

import conformsTo from 'lodash/conformsTo'
import defaults from 'lodash/defaults'
import isFunction from 'lodash/isFunction'
import isString from 'lodash/isString'
import isUndefined from 'lodash/isUndefined'

import { ApiOptions } from '../api-types'

export const apiOptions = (options: Partial<ApiOptions>): ApiOptions => {
  const shape = {
    token: isString,
    clientId: isString,
    onAuthenticationFailure: (cb: unknown) => isFunction(cb) || isUndefined(cb),
  }

  options = defaults(options, {
    clientId: undefined,
    onAuthenticationFailure: undefined,
  })

  invariant(conformsTo(options, shape), 'Expected valid options')

  return options as ApiOptions
}
