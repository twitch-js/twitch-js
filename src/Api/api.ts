import includes from 'lodash/includes'
import toUpper from 'lodash/toUpper'

import { ApiValidateResponse } from '../twitch'

import createLogger, { Logger } from '../utils/logger'
import fetchUtil, { FetchError } from '../utils/fetch'
import { AuthenticationError } from '../utils/error'

import * as validators from './utils/api-validators'

import { ApiOptions, ApiReadyStates, ApiFetchOptions } from './api-types'

/**
 * Make requests to Twitch API.
 *
 * ## Initializing
 *
 * ```js
 * const token = 'cfabdegwdoklmawdzdo98xt2fo512y'
 * const clientId = 'uo6dggojyb8d6soh92zknwmi5ej1q2'
 * const { api } = new TwitchJs({ token, clientId })
 * ```
 *
 * ## Making requests
 *
 * By default, the API client makes requests to the
 * [Helix API](https://dev.twitch.tv/docs/api), and exposes [[Api.get]],
 * [[Api.post]], [[Api.put]] and [[Api.delete]] methods. Query and body parameters are provided
 * via `options.search` and `options.body` properties, respectively.
 *
 * ### Examples
 *
 * #### Get bits leaderboard
 * ```js
 * api
 *   .get('bits/leaderboard', { search: { user_id: '44322889' } })
 *   .then(response => {
 *     // Do stuff with response ...
 *   })
 * ```
 *
 * #### Get the latest Overwatch live streams
 * ```
 * api
 *   .get('streams', { search: { game_id: '1234' } })
 *   .then(response => {
 *     // Do stuff with response ...
 *   })
 * ```
 *
 * #### Start a channel commercial
 * ```
 * api
 *   .post('/channels/commercial', {
 *     body: { broadcaster_id: '44322889', length: 30 },
 *   })
 *   .then(response => {
 *     // Do stuff with response ...
 *   })
 * ```
 */

class Api {
  private _options: ApiOptions
  private _log: Logger

  private _readyState: ApiReadyStates = ApiReadyStates.READY

  private _status!: ApiValidateResponse

  constructor(options: Partial<ApiOptions>) {
    this._options = validators.apiOptions(options)

    this._log = createLogger({ name: 'Api', ...this._options.log })
  }

  get readyState() {
    return this._readyState
  }

  get status() {
    return this._status
  }

  /**
   * Update client options.
   */
  updateOptions(options: Partial<ApiOptions>) {
    this._options = validators.apiOptions({ ...this._options, ...options })
  }

  /**
   * Initialize API client and retrieve status.
   * @see https://dev.twitch.tv/docs/v5/#root-url
   */
  async initialize(newOptions?: Partial<ApiOptions>) {
    if (newOptions) {
      this._options = validators.apiOptions({ ...this._options, ...newOptions })
    }

    if (!newOptions && this.readyState === 2) {
      return Promise.resolve()
    }

    const response = await fetchUtil<ApiValidateResponse>(
      'https://id.twitch.tv/oauth2/validate',
      { headers: { Authorization: `Bearer ${this._options.token}` } },
    )

    this._readyState = ApiReadyStates.INITIALIZED
    this._status = response
  }

  /**
   * Check if current credentials include `scope`.
   * @see https://dev.twitch.tv/docs/authentication/#twitch-api-v5
   */
  hasScope(
    /** Scope to check */
    scope: string,
  ): Promise<boolean> {
    return new Promise((resolve, reject) =>
      this.status?.scopes?.includes(scope) ? resolve(true) : reject(false),
    )
  }

  /**
   * GET endpoint.
   *
   * @example <caption>Get user follows (Helix)</caption>
   * ```
   * api.get('users/follows', { search: { to_id: '23161357' } })
   *   .then(response => {
   *     // Do stuff with response ...
   *   })
   * ```
   */
  get<T = any>(endpoint = '', options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, options)
  }

  /**
   * POST endpoint.
   */
  post<T = any>(endpoint: string, options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'post' })
  }

  /**
   * PUT endpoint.
   */
  put<T = any>(endpoint: string, options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'put' })
  }

  /**
   * DELETE endpoint.
   */
  delete<T = any>(endpoint: string, options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'delete' })
  }

  private _getAuthenticationHeaders(): RequestInit['headers'] {
    const { clientId, token } = this._options

    return {
      Authorization: `Bearer ${token}`,
      'Client-Id': clientId,
    }
  }

  private async _handleFetch<T = any>(
    maybeUrl = '',
    options: ApiFetchOptions = {},
  ) {
    const baseUrl = 'https://api.twitch.tv/helix'

    const url = `${baseUrl}/${maybeUrl}`

    const message = `${toUpper(options.method) || 'GET'} ${url}`

    const fetchProfiler = this._log.profile()

    const performRequest = async () => {
      const authenticationHeaders = this._getAuthenticationHeaders()

      const fetchOptions = {
        ...options,
        headers: {
          ...options.headers,
          ...authenticationHeaders,
        },
      }

      try {
        return await fetchUtil<T>(url, fetchOptions)
      } catch (error) {
        if (error instanceof FetchError && error.body.status === 401) {
          throw new AuthenticationError(error.message, error.body)
        }
        throw error
      }
    }

    let caughtError
    try {
      return await performRequest()
    } catch (error) {
      if (
        typeof this._options.onAuthenticationFailure === 'function' &&
        error instanceof AuthenticationError
      ) {
        const token = await this._handleAuthenticationFailure(error)

        if (token) {
          this._log.info(`${message} ... retrying with new token`)
          this.updateOptions({ token })
          return await performRequest()
        }
      }

      caughtError = error
      throw caughtError
    } finally {
      fetchProfiler.done(message, caughtError)
    }
  }

  private async _handleAuthenticationFailure(originError: Error) {
    try {
      return await this._options.onAuthenticationFailure?.()
    } catch (error) {
      this._log.error(error as Error, 'onAuthenticationFailure error occurred')
      throw originError
    }
  }
}

export default Api
