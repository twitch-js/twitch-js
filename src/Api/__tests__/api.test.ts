import { mocked } from 'ts-jest/utils'
import nodeFetch from 'cross-fetch'
import omit from 'lodash/omit'

import { FetchError } from '../../utils/fetch'

import Api from '../api'
import { AuthenticationError } from '../../utils/error'

jest.mock('cross-fetch')

const fetchUtil = mocked(nodeFetch, true)

describe('Api', () => {
  jest.setTimeout(500)

  const options = {
    token: 'TOKEN',
    clientId: 'CLIENT_ID',
    log: { enabled: false },
  }

  const fetchOptions = { search: { a: 'b' } }

  afterEach(() => {
    fetchUtil.mockClear()
  })

  test('should allow options to be updated', () => {
    const api = new Api(options)

    const log = { level: 'info' }

    const nextOptions = {
      clientId: 'NEXT_CLIENT_ID',
      token: 'NEXT_TOKEN',
      log,
    }

    api.updateOptions(nextOptions)

    expect(
      // @ts-expect-error private property
      api._options,
    ).toMatchObject({ ...nextOptions, log })
  })

  describe('initialize', () => {
    test('should set readyState and status', async () => {
      const api = new Api(options)
      await api.initialize()

      expect(api.readyState).toBe(2)
      expect(api.status).toMatchSnapshot()
    })

    test('should resolve if already initialized', async () => {
      const api = new Api(options)
      await api.initialize()
      const actual = await api.initialize()

      expect(actual).toMatchSnapshot()
    })

    test('should update client options', async () => {
      const api = new Api(options)
      expect(
        // @ts-expect-error private property
        api._options.token,
      ).toBe(options.token)

      const newToken = 'NEW_TOKEN'
      await api.initialize({ token: newToken })

      expect(
        // @ts-expect-error private property
        api._options.token,
      ).toBe(newToken)
    })
  })

  test('should create headers', async () => {
    const api = new Api(options)
    await api.get('')

    expect(fetchUtil.mock.calls).toMatchSnapshot()
  })

  describe('hasScope', () => {
    test('should reject if instance is uninitialized', async () => {
      const api = new Api(options)
      await expect(api.hasScope('user_read')).rejects.toBe(false)
    })

    test('should reject if scope is absent', async () => {
      const api = new Api(options)
      await api.initialize()
      await expect(api.hasScope('channel_read')).rejects.toBe(false)
    })

    test('should resolve if scope is present', async () => {
      const api = new Api(options)
      await api.initialize()
      await expect(api.hasScope('user_read')).resolves.toBe(true)
    })
  })

  describe('methods', () => {
    test('get should call fetch', async () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      await api.get(endpoint, fetchOptions)

      const [actualEndpoint, actualOpts] = fetchUtil.mock.calls[0]

      expect(actualEndpoint).toEqual(expect.stringContaining(endpoint))
      expect(actualOpts).toMatchObject(omit(fetchOptions, 'search'))
    })

    test('post should call fetch with method=post', async () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      const body = { broadcaster_id: 'BROADCASTER_ID' }
      await api.post(endpoint, { ...fetchOptions, body })

      const [, actualOpts] = fetchUtil.mock.calls[0]

      expect(actualOpts).toMatchSnapshot()
    })

    test('put should call fetch with method=put', async () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      await api.put(endpoint, fetchOptions)

      const [, actualOpts] = fetchUtil.mock.calls[0]

      expect(actualOpts).toMatchObject({ method: 'put' })
    })

    test('should throw on failure', async () => {
      expect.assertions(2)

      const api = new Api(options)

      try {
        await api.get('404')
      } catch (error) {
        if (error instanceof FetchError) {
          expect(error).toBeInstanceOf(FetchError)
          expect(error.body).toMatchSnapshot()
        }
      }
    })
  })

  describe('onAuthenticationFailure', () => {
    test('should throw AuthenticationError', async () => {
      const api = new Api({
        ...options,
        token: 'INVALID_TOKEN',
      })

      try {
        await api.get('401')
      } catch (error) {
        expect(error).toBeInstanceOf(AuthenticationError)
      }
    })

    test('should call onAuthenticationFailure', async () => {
      const onAuthenticationFailure = jest.fn(() => Promise.reject())
      const api = new Api({
        ...options,
        token: 'INVALID_TOKEN',
        onAuthenticationFailure,
      })

      try {
        await api.get('401')
      } catch (error) {
        expect(onAuthenticationFailure).toHaveBeenCalled()
        expect(error).toBeInstanceOf(AuthenticationError)
      }
    })

    test('should refresh token and try again', async () => {
      const onAuthenticationFailure = jest.fn(() => Promise.resolve('TOKEN'))
      const api = new Api({
        ...options,
        token: 'INVALID_TOKEN',
        onAuthenticationFailure,
      })

      try {
        await api.get('401')
      } catch (error) {
        expect(onAuthenticationFailure).toHaveBeenCalled()
        expect(fetchUtil).toHaveBeenCalledTimes(2)
        expect(
          // @ts-expect-error private property
          api._options.token,
        ).toEqual('TOKEN')
      }
    })
  })
})
