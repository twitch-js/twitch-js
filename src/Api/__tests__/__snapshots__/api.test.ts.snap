// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api initialize should resolve if already initialized 1`] = `undefined`;

exports[`Api initialize should set readyState and status 1`] = `
Object {
  "clientId": "wbmytr93xzw8zbg0p1izqyzzc5mbiz",
  "expiresIn": 5520838,
  "login": "twitchdev",
  "scopes": Array [
    "channel:read:subscriptions",
  ],
  "userId": "141981764",
}
`;

exports[`Api methods post should call fetch with method=post 1`] = `
Object {
  "body": "{\\"broadcaster_id\\":\\"BROADCASTER_ID\\"}",
  "headers": Object {
    "Authorization": "Bearer TOKEN",
    "Client-Id": "CLIENT_ID",
    "Content-Type": "application/json",
  },
  "method": "post",
}
`;

exports[`Api methods should throw on failure 1`] = `
Object {
  "error": "Not Found",
  "message": "The channel could not be found",
  "status": 404,
}
`;

exports[`Api should create headers 1`] = `
Array [
  Array [
    "https://api.twitch.tv/helix/",
    Object {
      "headers": Object {
        "Authorization": "Bearer TOKEN",
        "Client-Id": "CLIENT_ID",
      },
    },
  ],
]
`;
